fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -5087,3 +5088,606 @@ Future<Void> DDTeamCollection::run(Reference<DDTeamCollection> teamCollection,
+	collection->server_info[UID(1, 0)]->teams[0]->setHealthy(false);
+
+	/*
+	 * If the only available team is one where at least one server is low on
+	 * space, decline to pick that team. Every server must have some minimum
+	 * free space defined by the MIN_AVAILABLE_SPACE server knob.
+	 */
+
+	bool wantsNewServers = true;
+	bool wantsTrueBest = true;
+	bool preferLowerUtilization = true;
+	bool teamMustHaveShards = false;
+	std::vector<UID> completeSources{ UID(1, 0), UID(2, 0), UID(3, 0) };
+
+	state GetTeamRequest req(wantsNewServers, wantsTrueBest, preferLowerUtilization, teamMustHaveShards);
+	req.completeSources = completeSources;
+
+	Future<Void> a = collection->getTeam(req);
+	wait(a);
+
+	std::pair<Optional<Reference<IDataDistributionTeam>>, bool> resTeam = req.reply.getFuture().get();
+
+	ASSERT(!resTeam.first.present());
+
+	return Void();
+}
+
+TEST_CASE("/DataDistribution/GetTeam/ServerUtilizationNearCutoff") {
+	wait(Future<Void>(Void()));
+
+	Reference<IReplicationPolicy> policy =
+	    Reference<IReplicationPolicy>(new PolicyAcross(3, "zoneid", Reference<IReplicationPolicy>(new PolicyOne())));
+	state int processSize = 5;
+	state int teamSize = 3;
+	state std::unique_ptr<DDTeamCollection> collection = testTeamCollection(teamSize, policy, processSize);
+
+	GetStorageMetricsReply low_avail;
+	if (SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO > 0) {
+		/* Pick a capacity where MIN_AVAILABLE_SPACE_RATIO of the capacity would be higher than MIN_AVAILABLE_SPACE */
+		low_avail.capacity.bytes = SERVER_KNOBS->MIN_AVAILABLE_SPACE * (2 / SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO);
+	} else {
+		low_avail.capacity.bytes = 2000 * 1024 * 1024;
+	}
+	low_avail.available.bytes = (SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO * 1.1) * low_avail.capacity.bytes;
+	low_avail.load.bytes = 90 * 1024 * 1024;
+
+	GetStorageMetricsReply high_avail;
+	high_avail.capacity.bytes = 2000 * 1024 * 1024;
+	high_avail.available.bytes = 800 * 1024 * 1024;
+	high_avail.load.bytes = 90 * 1024 * 1024;
+
+	collection->addTeam(std::set<UID>({ UID(1, 0), UID(2, 0), UID(3, 0) }), true);
+	collection->addTeam(std::set<UID>({ UID(2, 0), UID(3, 0), UID(4, 0) }), true);
+	collection->addTeam(std::set<UID>({ UID(3, 0), UID(4, 0), UID(5, 0) }), true);
+	collection->enableBuildingTeams(false);
+	collection->clearCheckTeamDelay();
+
+	collection->server_info[UID(1, 0)]->serverMetrics = high_avail;
+	collection->server_info[UID(2, 0)]->serverMetrics = low_avail;
+	collection->server_info[UID(3, 0)]->serverMetrics = high_avail;
+	collection->server_info[UID(4, 0)]->serverMetrics = low_avail;
+	collection->server_info[UID(5, 0)]->serverMetrics = high_avail;
+	collection->server_info[UID(1, 0)]->teams[0]->setHealthy(false);
+
+	/*
+	 * If the only available team is one where all servers are low on space,
+	 * test that each server has at least MIN_AVAILABLE_SPACE_RATIO (server knob)
+	 * percentage points of capacity free before picking that team.
+	 */
```

Comment:
We should have another knob to ensure that we don't pick a team if it is too close to the cutoff.

Reply:
I have that coming in a separate PR - the one that resolves FDB#4271. The test case result will get updated slightly then.